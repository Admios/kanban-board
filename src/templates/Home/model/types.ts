export interface Todo {
  id: string;
  stateId: string;
  text: string;
  position: number;
  owner: string;
}

export interface State {
  id: string;
  name: string;
  position: number;
  color: string;
  owner: string;
}
