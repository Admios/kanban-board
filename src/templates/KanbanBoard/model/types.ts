export interface Task {
  id: string;
  stateId: string;
  text: string;
  position: number;
}

export interface State {
  id: string;
  boardId: string;
  name: string;
  position: number;
  color: string;
}

export interface Board {
  id: string;
  name: string;
  owner: string;
}
