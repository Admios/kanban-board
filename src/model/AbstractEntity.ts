/**
 * Entities represent how the data is stored in the database.
 */
export interface AbstractEntity extends Record<string, any> {}
