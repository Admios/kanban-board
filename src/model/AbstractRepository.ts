export interface AbstractRepository<T> {
  findById(id: string): Promise<T>;
  list(): Promise<T[]>;
  create(entity: T): Promise<T>;
  update(entity: T): Promise<T>;
  delete(entity: T): Promise<T>;
}
