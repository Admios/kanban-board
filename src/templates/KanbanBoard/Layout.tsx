import { AddStateModal } from "./AddStateModal";
import { AddTaskModal } from "./AddTaskModal";
import { EditTaskModal } from "./EditTaskModal";
import { EmptyView } from "./EmptyView";
import { KanbanColumnList } from "./KanbanColumnList";
import styles from "./Layout.module.scss";
import { Navbar } from "./Navbar";
import { useModalState } from "./useModalState";

export function Layout() {
  const [modals, dispatch] = useModalState();

  return (
    <section className={styles.container}>
      <Navbar onOpenStateDialog={() => dispatch({ key: "ADD_STATE::OPEN" })} />

      <div className={styles.content}>
        <KanbanColumnList
          onOpenEditTaskModal={(task) =>
            dispatch({ key: "EDIT_TASK::OPEN", payload: { task } })
          }
          onOpenCreateTaskModal={(stateId) =>
            dispatch({ key: "ADD_TASK::OPEN", payload: { stateId } })
          }
        >
          <EmptyView
            onOpenStateDialog={() => dispatch({ key: "ADD_STATE::OPEN" })}
            isStateDialogOpen={modals.addStateIsOpen}
          />
        </KanbanColumnList>
      </div>

      <AddStateModal
        isOpen={modals.addStateIsOpen}
        onClose={() => dispatch({ key: "ADD_STATE::CLOSE" })}
      />
      <AddTaskModal
        isOpen={!!modals.editingStateId}
        onClose={() => dispatch({ key: "ADD_TASK::CLOSE" })}
        stateId={modals.editingStateId}
      />
      <EditTaskModal
        isOpen={!!modals.editingTask}
        onClose={() => dispatch({ key: "EDIT_TASK::CLOSE" })}
        task={modals.editingTask}
      />
    </section>
  );
}
