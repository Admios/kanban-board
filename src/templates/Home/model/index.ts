import { StateDTO } from "@/model/State";
import { TaskDTO } from "@/model/Task";
import { UserDTO } from "@/model/User";
import { Immutable, produce } from "immer";
import { StateCreator, create } from "zustand";
import { State, Task } from "./types";

export type { State, Task } from "./types";

type HomeState = Immutable<{
  tasks: Record<string, Task[]>;
  states: Record<string, State>;
  user?: UserDTO;
}>;

interface HomeActions {
  initialize(
    initialTasks: TaskDTO[],
    initialStates: StateDTO[],
    initialUser?: UserDTO,
  ): void;
  addTask(newTask: Task): Task;
  moveTask(
    newTask: Task,
    fromStateId: string,
    toStateId: string,
    position: number,
  ): void;
  addState(newState: State): State;
  editTask(id: string, updatedValues: Partial<Task>): void;
  deleteTask: (id: string) => void;
}

const stateCreator: StateCreator<HomeState & HomeActions> = (set, get) => ({
  tasks: {},
  states: {},

  initialize(initialTasks, initialStates, initialUser) {
    const statesMap: Record<string, State> = {};
    const tasksMap: Record<string, Task[]> = {};

    initialStates.forEach((backendState) => {
      const result: State = {
        ...backendState,
        id: backendState.id,
        position: backendState.position,
      };
      statesMap[result.id] = result;
    });

    initialTasks.forEach((backendTask) => {
      const { stateId } = backendTask;
      if (!tasksMap[stateId]) {
        tasksMap[stateId] = [];
      }
      tasksMap[stateId][backendTask.position] = {
        ...backendTask,
        stateId,
        id: backendTask.id,
        position: backendTask.position,
      };
    });

    set({ tasks: tasksMap, states: statesMap, user: initialUser });
  },

  addTask: (newTask) => {
    const result = { ...newTask };
    const tasks = produce(get().tasks, (draft) => {
      if (!draft[newTask.stateId]) {
        draft[newTask.stateId] = [];
      }

      draft[newTask.stateId].push(result);
    });

    set({ tasks });
    return result;
  },

  moveTask: (newTask, fromStateId, toStateId, position) => {
    const tasks = produce(get().tasks, (draftState) => {
      const sourceState = (draftState[fromStateId] ?? []).filter(
        (task) => task.id !== newTask.id,
      );

      let destinationState = draftState[toStateId] ?? [];
      if (fromStateId === toStateId) {
        destinationState = sourceState;
      }

      if (position <= destinationState.length) {
        destinationState.splice(position, 0, newTask);
      } else {
        destinationState.push(newTask);
      }

      draftState[fromStateId] = sourceState;
      draftState[toStateId] = destinationState;
    });

    set({ tasks });
  },

  addState: (newState) => {
    const currentModel = get();
    const tasks = produce(currentModel.tasks, (draft) => {
      draft[newState.id] = [];
    });
    const states = produce(currentModel.states, (draft) => {
      const position = Object.keys(currentModel.states).length;
      draft[newState.id] = { ...newState, position };
    });

    set({ states, tasks });
    return states[newState.id];
  },

  editTask: (taskId, updatedValues) => {
    const currentState = get();
    const tasks = produce(currentState.tasks, (draft) => {
      for (const stateId in draft) {
        const stateTasks = draft[stateId];
        const taskIndex = stateTasks.findIndex((task) => task.id === taskId);
        if (taskIndex > -1) {
          stateTasks[taskIndex] = {
            ...stateTasks[taskIndex],
            ...updatedValues,
          };
          break;
        }
      }
    });

    set({ tasks });
  },

  deleteTask: (taskId) => {
    const currentState = get();
    const tasks = produce(currentState.tasks, (draft) => {
      for (const stateId in draft) {
        const stateTasks = draft[stateId];
        const taskIndex = stateTasks.findIndex((task) => task?.id === taskId);
        if (taskIndex > -1) {
          stateTasks.splice(taskIndex, 1);
          break;
        }
      }
    });

    set({ tasks });
  },
});

export const useZustand = create(stateCreator);
