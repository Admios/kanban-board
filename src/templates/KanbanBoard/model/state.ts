import { StateDTO } from "@/model/State";
import { TaskDTO } from "@/model/Task";
import { UserDTO } from "@/model/User";
import { Immutable } from "immer";
import { State, Task } from "./types";

export type KanbanState = Immutable<{
  /**
   * The order of the states in the board.
   */
  statesOrder: string[];

  /**
   * The order of the tasks in each state.
   */
  tasksOrder: Record<string, string[]>;

  /**
   * Key: taskId. Value: Task.
   */
  tasks: Record<string, Task>;

  /**
   * Key: stateId. Value: State.
   */
  states: Record<string, State>;

  /**
   * The user that is currently logged in
   */
  user?: UserDTO;
}>;

export interface KanbanActions {
  setUser(user?: UserDTO): void;
  initialize(initialTasks: TaskDTO[], initialStates: StateDTO[]): void;
  addTask(newTask: Task): Task;
  moveTask(
    newTask: Task,
    fromStateId: string,
    toStateId: string,
    position: number,
  ): void;
  addState(newState: State): State;
  editTask(id: string, updatedValues: Partial<Task>): void;
  deleteTask: (id: string) => void;

  clear(): void;
}
