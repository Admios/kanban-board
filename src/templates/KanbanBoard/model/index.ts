import { produce } from "immer";
import { StateCreator, create } from "zustand";
import { State, Task } from "./types";
import { KanbanState, KanbanActions } from "./state";

export type { State, Task } from "./types";

const stateCreator: StateCreator<KanbanState & KanbanActions> = (set, get) => ({
  statesOrder: [],
  tasksOrder: {},
  tasks: {},
  states: {},

  initialize(initialTasks, initialStates) {
    const statesOrderSparseArray: string[] = [];
    const tasksOrder: Record<string, string[]> = {};
    const states: Record<string, State> = {};
    const tasks: Record<string, Task> = {};

    initialTasks.forEach((backendTask) => {
      tasks[backendTask.id] = backendTask;
      if (!tasksOrder[backendTask.stateId]) {
        tasksOrder[backendTask.stateId] = [];
      }
      // Warning: saved as sparse array! Compression is done later
      tasksOrder[backendTask.stateId][backendTask.position] = backendTask.id;
    });

    initialStates.forEach((backendState) => {
      states[backendState.id] = backendState;
      // Warning: saved as sparse array! compressed later
      statesOrderSparseArray[backendState.position] = backendState.id;
      // Compress sparse array
      tasksOrder[backendState.id] = (tasksOrder[backendState.id] ?? []).filter(
        (id) => !!id,
      );
    });

    // Compress sparse array
    const statesOrder = statesOrderSparseArray.filter((id) => !!id);

    set({
      statesOrder,
      tasksOrder,
      tasks,
      states,
    });
  },

  setUser(user) {
    set({ user });
  },

  addTask: (newTask) => {
    const tasks = produce(get().tasks, (draft) => {
      draft[newTask.stateId] = newTask;
    });

    const tasksOrder = produce(get().tasksOrder, (draft) => {
      if (!draft[newTask.stateId]) {
        draft[newTask.stateId] = [];
      }
      draft[newTask.stateId].push(newTask.id);
    });

    set({ tasks, tasksOrder });
    return newTask;
  },

  moveTask: (newTask, fromStateId, toStateId, position) => {
    const newTasksOrder = produce(get().tasksOrder, (draftState) => {
      const sourceState = (draftState[fromStateId] ?? []).filter(
        (task) => task !== newTask.id,
      );

      let destinationState = draftState[toStateId] ?? [];
      if (fromStateId === toStateId) {
        destinationState = sourceState;
      }

      if (position <= destinationState.length) {
        destinationState.splice(position, 0, newTask.id);
      } else {
        destinationState.push(newTask.id);
      }

      draftState[fromStateId] = sourceState;
      draftState[toStateId] = destinationState;
    });

    set({ tasksOrder: newTasksOrder });
  },

  addState: (newState) => {
    const tasksOrder = produce(get().tasksOrder, (draft) => {
      draft[newState.id] = [];
    });

    const statesOrder = produce(get().statesOrder, (draft) => {
      draft.push(newState.id);
    });

    const states = produce(get().states, (draft) => {
      draft[newState.id] = newState;
    });

    set({ states, tasksOrder, statesOrder });
    return states[newState.id];
  },

  editTask: (taskId, updatedValues) => {
    const tasks = produce(get().tasks, (draft) => {
      draft[taskId] = { ...draft[taskId], ...updatedValues };
    });

    set({ tasks });
  },

  deleteTask: (taskId) => {
    const { stateId } = get().tasks[taskId];
    const tasksOrder = produce(get().tasksOrder, (draft) => {
      draft[stateId] = draft[stateId].filter((id) => id !== taskId);
    });

    const tasks = produce(get().tasks, (draft) => {
      delete draft[taskId];
    });

    set({ tasks, tasksOrder });
  },
});

export const useZustand = create(stateCreator);
