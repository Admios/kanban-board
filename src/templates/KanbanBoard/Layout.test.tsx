import { act, render, screen, waitFor, within } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { Layout } from "./Layout";
import { useZustand } from "./model";

jest.mock("./clearCookies.ts");
jest.mock("./kanbanActions.ts");

afterEach(() => {
  useZustand.setState({
    states: {},
    tasks: {},
    user: {
      username: "test",
    },
  });
});

function setInitialState() {
  // NOTE: The state's positions are not in order!
  // They should get sorted by the component itself
  act(() => {
    useZustand.setState({
      states: {
        "state-1": {
          id: "state-1",
          name: "To do",
          position: 100,
          color: "red",
          owner: "test",
        },
        "state-2": {
          id: "state-2",
          name: "In progress",
          position: 0,
          color: "blue",
          owner: "test",
        },
        "state-3": {
          id: "state-3",
          name: "Done",
          position: 3,
          color: "green",
          owner: "test",
        },
      },
      tasks: {
        "state-1": [],
        "state-2": [],
        "state-3": [],
      },
    });
  });
}

it("should render the sorted states", () => {
  setInitialState();

  const { container } = render(<Layout />);
  expect(container).toMatchSnapshot("Default Kanban Page");

  ["In progress", "To do", "Done"].forEach((stateTitle) => {
    const element = screen.queryByText(stateTitle);
    expect(element).toBeInTheDocument();
  });
});

// it should render when empty
it("should render when empty", () => {
  const { container } = render(<Layout />);
  expect(container).toMatchSnapshot("Empty Kanban Page");

  ["In progress", "To do", "Done"].forEach((stateTitle) => {
    const element = screen.queryByText(stateTitle);
    expect(element).not.toBeInTheDocument();
  });
});

it("should open the AddStateModal when button is pressed", async () => {
  render(<Layout />);

  // Click the button
  const button = await screen.findByTestId("add-state-button")
  expect(button).toBeInTheDocument();

  await userEvent.click(button);

  const modal = screen.getByRole("dialog");
  modal.setAttribute("style", ""); // This value has animation! We don't want that in our snapshot
  expect(modal).toMatchSnapshot("AddStateModal");
});

describe("sould create 10 random tasks when button is pressed", () => {
  it("should add new tasks on the first state", async () => {
    setInitialState();
    render(<Layout />);

    const createRandomTasksButton = await screen.findByTestId("create-random-task-button")
    expect(createRandomTasksButton).toBeInTheDocument();

    const state = screen.getByTitle("In progress");
    expect(state).toBeInTheDocument();

    expect(within(state).queryAllByTitle("task")).toHaveLength(0);

    await userEvent.click(createRandomTasksButton);

    await waitFor(() =>
      expect(within(state).getAllByTitle("task")).toHaveLength(10),
    );
  });

  it("should add new tasks on a new state", async () => {
    render(<Layout />);

    const createRandomTasksButton = await screen.findByTestId("create-random-task-button")
    expect(createRandomTasksButton).toBeInTheDocument();

    let randomState: HTMLElement | null;

    randomState = screen.queryByTitle("Random State");
    expect(randomState).not.toBeInTheDocument();

    await userEvent.click(createRandomTasksButton);

    randomState = screen.getByTitle("Random State");
    await waitFor(() => expect(randomState).toBeInTheDocument());
    expect(within(randomState).getAllByTitle("task")).toHaveLength(10);
  });
});
