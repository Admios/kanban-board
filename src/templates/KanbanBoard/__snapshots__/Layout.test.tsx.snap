// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should open the AddStateModal when button is pressed: AddStateModal 1`] = `
<div
  aria-labelledby="add-state-modal-title"
  class="modal is-active"
  role="dialog"
>
  <div
    class="modal-background"
  />
  <div
    class="modal-card"
  >
    <header
      class="modal-card-head"
    >
      <p
        class="modal-card-title"
        id="add-state-modal-title"
      >
        Add State
      </p>
      <button
        aria-label="close"
        class="delete"
      />
    </header>
    <section
      class="modal-card-body"
    >
      <div
        class="field"
      >
        <label
          class="label"
        >
          Name:
        </label>
        <div
          class="control"
        >
          <input
            class="input is-danger"
            type="text"
            value=""
          />
        </div>
        <p
          class="help is-danger"
        >
          Name is required.
        </p>
      </div>
    </section>
    <footer
      class="modal-card-foot"
    >
      <div
        class="buttons"
      >
        <button
          class="button"
        >
          Close
        </button>
        <button
          class="button is-success"
          disabled=""
        >
          Add State
        </button>
      </div>
    </footer>
  </div>
</div>
`;

exports[`should render the sorted states: Default Kanban Page 1`] = `
<div>
  <section
    class="container"
  >
    <nav
      aria-label="main navigation"
      class="navbar"
      role="navigation"
    >
      <div
        class="navbar-brand"
      >
        <a
          class="navbar-item"
          href="/"
        >
          <img
            alt="Logo"
            data-nimg="1"
            decoding="async"
            height="28"
            loading="lazy"
            src="/admios-logo.svg"
            style="color: transparent;"
            width="112"
          />
        </a>
        <a
          aria-expanded="false"
          aria-label="menu"
          class="navbar-burger"
          data-target="main-navbar-inner"
          role="button"
        >
          <span
            aria-hidden="true"
          />
          <span
            aria-hidden="true"
          />
          <span
            aria-hidden="true"
          />
        </a>
      </div>
      <div
        class="navbar-menu"
        id="main-navbar-inner"
      >
        <div
          class="navbar-start"
        >
          <a
            class="navbar-item"
            href="/"
          >
            My Boards
          </a>
          <a
            class="navbar-item"
          >
            Add State
          </a>
        </div>
        <div
          class="navbar-end"
        >
          <div
            class="navbar-item"
          >
            Username: 
            test
          </div>
          <a
            class="navbar-item"
          >
            Logout
          </a>
        </div>
      </div>
    </nav>
    <div
      class="columnList content"
    >
      <section
        class="card container"
        style="--column-color: blue;"
      >
        <header
          class="card-header"
        >
          <p
            class="card-header-title"
          >
            In progress
          </p>
        </header>
        <div
          class="card-content cardContent"
        />
        <footer
          class="card-footer"
        >
          <a
            class="card-footer-item"
          >
            Add task
          </a>
        </footer>
      </section>
      <section
        class="card container"
        style="--column-color: green;"
      >
        <header
          class="card-header"
        >
          <p
            class="card-header-title"
          >
            Done
          </p>
        </header>
        <div
          class="card-content cardContent"
        />
        <footer
          class="card-footer"
        >
          <a
            class="card-footer-item"
          >
            Add task
          </a>
        </footer>
      </section>
      <section
        class="card container"
        style="--column-color: red;"
      >
        <header
          class="card-header"
        >
          <p
            class="card-header-title"
          >
            To do
          </p>
        </header>
        <div
          class="card-content cardContent"
        />
        <footer
          class="card-footer"
        >
          <a
            class="card-footer-item"
          >
            Add task
          </a>
        </footer>
      </section>
    </div>
    <div
      aria-labelledby="add-state-modal-title"
      class="modal"
      role="dialog"
    >
      <div
        class="modal-background"
      />
      <div
        class="modal-card"
      >
        <header
          class="modal-card-head"
        >
          <p
            class="modal-card-title"
            id="add-state-modal-title"
          >
            Add State
          </p>
          <button
            aria-label="close"
            class="delete"
          />
        </header>
        <section
          class="modal-card-body"
        >
          <div
            class="field"
          >
            <label
              class="label"
            >
              Name:
            </label>
            <div
              class="control"
            >
              <input
                class="input is-danger"
                type="text"
                value=""
              />
            </div>
            <p
              class="help is-danger"
            >
              Name is required.
            </p>
          </div>
        </section>
        <footer
          class="modal-card-foot"
        >
          <div
            class="buttons"
          >
            <button
              class="button"
            >
              Close
            </button>
            <button
              class="button is-success"
              disabled=""
            >
              Add State
            </button>
          </div>
        </footer>
      </div>
    </div>
    <div
      aria-labelledby="add-task-modal-title"
      class="modal"
      role="dialog"
    >
      <div
        class="modal-background"
      />
      <div
        class="modal-card"
      >
        <header
          class="modal-card-head"
        >
          <p
            class="modal-card-title"
            id="add-task-modal-title"
          >
            Add Task
          </p>
          <button
            aria-label="close"
            class="delete"
          />
        </header>
        <section
          class="modal-card-body"
        >
          <div
            class="field"
          >
            <label
              class="label"
            >
              Name:
            </label>
            <div
              class="control"
            >
              <input
                class="input"
                type="text"
                value=""
              />
            </div>
            <p
              class="help is-danger"
            >
              Task is required.
            </p>
          </div>
        </section>
        <footer
          class="modal-card-foot"
        >
          <div
            class="buttons"
          >
            <button
              class="button"
            >
              Close
            </button>
            <button
              class="button is-success"
              disabled=""
            >
              Add Task
            </button>
          </div>
        </footer>
      </div>
    </div>
    <div
      aria-labelledby="edit-task-modal-title"
      class="modal"
      role="dialog"
    >
      <div
        class="modal-background"
      />
      <div
        class="modal-card"
      >
        <header
          class="modal-card-head"
        >
          <p
            class="modal-card-title"
            id="edit-taskmodal-title"
          >
            Edit Task
          </p>
          <button
            aria-label="close"
            class="delete"
          />
        </header>
        <section
          class="modal-card-body"
        >
          <div
            class="field"
          >
            <label
              class="label"
            >
              Name:
            </label>
            <div
              class="control"
            >
              <input
                class="input"
                type="text"
                value=""
              />
            </div>
            <p
              class="help is-danger"
            >
              Text is required.
            </p>
          </div>
        </section>
        <footer
          class="modal-card-foot"
        >
          <div
            class="buttons"
          >
            <button
              class="button"
            >
              Close
            </button>
            <button
              class="button is-success"
              disabled=""
            >
              Edit Task
            </button>
          </div>
        </footer>
      </div>
    </div>
  </section>
</div>
`;

exports[`should render when empty: Empty Kanban Page 1`] = `
<div>
  <section
    class="container"
  >
    <nav
      aria-label="main navigation"
      class="navbar"
      role="navigation"
    >
      <div
        class="navbar-brand"
      >
        <a
          class="navbar-item"
          href="/"
        >
          <img
            alt="Logo"
            data-nimg="1"
            decoding="async"
            height="28"
            loading="lazy"
            src="/admios-logo.svg"
            style="color: transparent;"
            width="112"
          />
        </a>
        <a
          aria-expanded="false"
          aria-label="menu"
          class="navbar-burger"
          data-target="main-navbar-inner"
          role="button"
        >
          <span
            aria-hidden="true"
          />
          <span
            aria-hidden="true"
          />
          <span
            aria-hidden="true"
          />
        </a>
      </div>
      <div
        class="navbar-menu"
        id="main-navbar-inner"
      >
        <div
          class="navbar-start"
        >
          <a
            class="navbar-item"
            href="/"
          >
            My Boards
          </a>
          <a
            class="navbar-item"
          >
            Add State
          </a>
        </div>
        <div
          class="navbar-end"
        >
          <div
            class="navbar-item"
          >
            Username: 
            test
          </div>
          <a
            class="navbar-item"
          >
            Logout
          </a>
        </div>
      </div>
    </nav>
    <div
      class="columnList content"
    >
      <div
        class="card emptyContent"
      >
        <header
          class="card-header"
        >
          <p
            class="card-header-title"
          >
            Your Workspace is empty!
          </p>
        </header>
        <div
          class="card-content"
        >
          <div
            class="content"
          >
            <p>
              You can create your first column below.
            </p>
          </div>
        </div>
        <footer
          class="card-footer"
        >
          <a
            class="card-footer-item"
          >
            Create Column
          </a>
        </footer>
      </div>
    </div>
    <div
      aria-labelledby="add-state-modal-title"
      class="modal"
      role="dialog"
    >
      <div
        class="modal-background"
      />
      <div
        class="modal-card"
      >
        <header
          class="modal-card-head"
        >
          <p
            class="modal-card-title"
            id="add-state-modal-title"
          >
            Add State
          </p>
          <button
            aria-label="close"
            class="delete"
          />
        </header>
        <section
          class="modal-card-body"
        >
          <div
            class="field"
          >
            <label
              class="label"
            >
              Name:
            </label>
            <div
              class="control"
            >
              <input
                class="input is-danger"
                type="text"
                value=""
              />
            </div>
            <p
              class="help is-danger"
            >
              Name is required.
            </p>
          </div>
        </section>
        <footer
          class="modal-card-foot"
        >
          <div
            class="buttons"
          >
            <button
              class="button"
            >
              Close
            </button>
            <button
              class="button is-success"
              disabled=""
            >
              Add State
            </button>
          </div>
        </footer>
      </div>
    </div>
    <div
      aria-labelledby="add-task-modal-title"
      class="modal"
      role="dialog"
    >
      <div
        class="modal-background"
      />
      <div
        class="modal-card"
      >
        <header
          class="modal-card-head"
        >
          <p
            class="modal-card-title"
            id="add-task-modal-title"
          >
            Add Task
          </p>
          <button
            aria-label="close"
            class="delete"
          />
        </header>
        <section
          class="modal-card-body"
        >
          <div
            class="field"
          >
            <label
              class="label"
            >
              Name:
            </label>
            <div
              class="control"
            >
              <input
                class="input"
                type="text"
                value=""
              />
            </div>
            <p
              class="help is-danger"
            >
              Task is required.
            </p>
          </div>
        </section>
        <footer
          class="modal-card-foot"
        >
          <div
            class="buttons"
          >
            <button
              class="button"
            >
              Close
            </button>
            <button
              class="button is-success"
              disabled=""
            >
              Add Task
            </button>
          </div>
        </footer>
      </div>
    </div>
    <div
      aria-labelledby="edit-task-modal-title"
      class="modal"
      role="dialog"
    >
      <div
        class="modal-background"
      />
      <div
        class="modal-card"
      >
        <header
          class="modal-card-head"
        >
          <p
            class="modal-card-title"
            id="edit-taskmodal-title"
          >
            Edit Task
          </p>
          <button
            aria-label="close"
            class="delete"
          />
        </header>
        <section
          class="modal-card-body"
        >
          <div
            class="field"
          >
            <label
              class="label"
            >
              Name:
            </label>
            <div
              class="control"
            >
              <input
                class="input"
                type="text"
                value=""
              />
            </div>
            <p
              class="help is-danger"
            >
              Text is required.
            </p>
          </div>
        </section>
        <footer
          class="modal-card-foot"
        >
          <div
            class="buttons"
          >
            <button
              class="button"
            >
              Close
            </button>
            <button
              class="button is-success"
              disabled=""
            >
              Edit Task
            </button>
          </div>
        </footer>
      </div>
    </div>
  </section>
</div>
`;
